src/core/storage/sframe_interface/unity_sframe.cpp

```diff
@@ -1200,6 +1202,78 @@ unity_sframe::random_split(float percent, uint64_t random_seed, bool exact) {
   return logical_filter_split(logical_filter_array);
 }
 
+std::shared_ptr<unity_sframe_base> unity_sframe::shuffle() {
+  log_func_entry();
+
+  std::vector<std::string> column_names = this->column_names();
+  const size_t num_buckets = (this->size() / SFRAME_SHUFFLE_BUCKET_SIZE) + 1;
+
+  // Create a column of random ints between 0 and (num_buckets - 1).
+  auto temp_groupby_column = std::static_pointer_cast<unity_sarray>(
+    unity_sarray::make_uniform_int_array(this->size(), num_buckets));
+  const std::string rand_int_column_name("Random Ints");
+  std::shared_ptr<unity_sframe> temp(new unity_sframe());
+  temp->add_column(temp_groupby_column, rand_int_column_name);
+
+  // Pack columns so we can group by concatenate
+  std::shared_ptr<unity_sarray_base> packed_columns = this->pack_columns(column_names,
+                                                                         column_names,
+                                                                         flex_type_enum::LIST,
+                                                                         turi::flex_undefined());
+  const std::string packed_data_column_name("Packed Data");
+  temp->add_column(packed_columns, packed_data_column_name);
+
+  // Group by concatenate on the random int column. This randomly bucketizes.
+  const std::string buckets_column_name("Buckets");
+  std::shared_ptr<unity_sframe_base> bucketized_sframe = temp->groupby_aggregate({rand_int_column_name},
+                                                                                 {{packed_data_column_name}},
+                                                                                 {buckets_column_name},
+                                                                                 {"__builtin__concat__list__"});
+  std::shared_ptr<unity_sarray_base> bucketized_sarray = bucketized_sframe->select_column(buckets_column_name);
+
+  // Shuffle each bucket
+  size_t num_threads = thread::cpu_count();
+  gl_sarray_writer writer(flex_type_enum::LIST, num_threads);
+  gl_sarray gl_bucketized_sarray = gl_sarray(bucketized_sarray);
+
+  in_parallel([&](size_t thread_id, size_t n_threads) {
+      size_t idx_start = (gl_bucketized_sarray.size() * thread_id) / n_threads;
+      size_t idx_end = (gl_bucketized_sarray.size() * (thread_id + 1) ) / n_threads;
+
+      gl_sarray_range ra = gl_bucketized_sarray.range_iterator(idx_start, idx_end);
+      auto cur_bucket = ra.begin();
+      unsigned int seed = static_cast<unsigned int>(random::pure_random_seed());
+      while (cur_bucket != ra.end()) {
+        // shuffle the indexes for the current bucket
+        auto indexes = std::vector<int>(cur_bucket->size());
+        std::iota(indexes.begin(), indexes.end(), 0);
+        std::shuffle(indexes.begin(), indexes.end(), std::default_random_engine(seed));
```

Comment:
nit: we can just do `auto rng = std::default_random_engine(seed);` once in each thread, then pass that into the std::shuffle here. Subsequent draws will still be random.